specs/light_client/merkle_proofs.md

```diff
@@ -64,43 +59,131 @@ y_data_root  len(y)
+
+
+def get_generalized_index(typ: Type, path: List[Union[int, str]]) -> GeneralizedIndex:
+    """
+    Converts a path (eg. `[7, "foo", 3]` for `x[7].foo[3]`, `[12, "bar", "__len__"]` for
+    `len(x[12].bar)`) into the generalized index representing its position in the Merkle tree.
+    """
+    root = 1
+    for p in path:
+        assert not issubclass(typ, BasicValue)  # If we descend to a basic type, the path cannot continue further
+        if p == '__len__':
+            typ, root = uint64, root * 2 + 1 if issubclass(typ, (List, Bytes)) else None
+        else:
+            pos, _, _ = get_item_position(typ, p)
+            root = root * (2 if issubclass(typ, (List, Bytes)) else 1) * next_power_of_two(chunk_count(typ)) + pos
+            typ = get_elem_type(typ, p)
+    return root
 ```
 
-We can now define a function `get_generalized_indices(object: Any, path: List[int], root: int=1) -> List[int]` that converts an object and a path to a set of generalized indices (note that for constant-sized objects, there is only one generalized index and it only depends on the path, but for dynamically sized objects the indices may depend on the object itself too). For dynamically-sized objects, the set of indices will have more than one member because of the need to access an array's length to determine the correct generalized index for some array access.
+### Helpers for generalized indices
+
+#### `concat_generalized_indices`
 
 ```python
-def get_generalized_indices(obj: Any, path: List[int], root: int=1) -> List[int]:
-    if len(path) == 0:
-        return [root]
-    elif isinstance(obj, Vector):
-        items_per_chunk = (32 // len(serialize(x))) if isinstance(x, int) else 1
-        new_root = root * next_power_of_2(len(obj) // items_per_chunk) + path[0] // items_per_chunk
-        return get_generalized_indices(obj[path[0]], path[1:], new_root)
-    elif isinstance(obj, List) and path[0] == LENGTH_FLAG:
-        return [root * 2 + 1]
-    elif isinstance(obj, List) and isinstance(path[0], int):
-        assert path[0] < len(obj)
-        items_per_chunk = (32 // len(serialize(x))) if isinstance(x, int) else 1
-        new_root = root * 2 * next_power_of_2(len(obj) // items_per_chunk) + path[0] // items_per_chunk
-        return [root *2 + 1] + get_generalized_indices(obj[path[0]], path[1:], new_root)
-    elif hasattr(obj, "fields"):
-        field = list(fields.keys())[path[0]]
-        new_root = root * next_power_of_2(len(fields)) + path[0]
-        return get_generalized_indices(getattr(obj, field), path[1:], new_root)
-    else:
-        raise Exception("Unknown type / path")
+def concat_generalized_indices(*indices: Sequence[GeneralizedIndex]) -> GeneralizedIndex:
+    """
+    Given generalized indices i1 for A -> B, i2 for B -> C .... i_n for Y -> Z, returns
+    the generalized index for A -> Z.
+    """
+    o = GeneralizedIndex(1)
+    for i in indices:
+        o = o * get_previous_power_of_2(i) + i
```

Comment:
Do I read this right? We do padding by the length that does not include the leading 1 bit, but not actually remove the bit from the index before appending the bits to the output?

Example of what it looks like is happening here (assuming `get_previous_power_of_2(i)` just returns the closest power of 2 at `i` or lower):
```txt
indices = [ # (in binary for example purposes)
0b10001
0b1011
0b110001
]
o = 1
o = 1 0000 # pad
o = 1 0 0001 # add
o =    1 0001 000 # pad
o =    1 0010 011 # add
o =    1 0001 011 00000 # pad
o =    1 0001 100 10001 # add
```
What I expected:
```txt
o = 1
o = 1 0000 # pad
o = 1 0001 # add
o = 1 0001 000 # pad
o = 1 0001 011 # add
o = 1 0001 011 00000 # pad
o = 1 0001 011 10001 # add
```

Reply:
Fixed! For inputs `(10001, 1011, 110001)` the output should indeed be `1 0001 011 10001` now.